### PR TITLE
Add MCP session state machine and request handling

### PR DESCRIPTION
**Pull request by sergey-shandar**
## Summary
Implements the core MCP (Model Context Protocol) session lifecycle and request handling logic, including state transitions, capability-gated method dispatch, and JSON-RPC error handling.

## Key Changes
- **Session state machine**: Added `McpSessionState` type representing two phases (`uninitialized` and `initialized`) with proper state transitions
- **Configuration and handlers**: Introduced `McpConfig` type for server configuration and `McpHandlers` for method implementations
- **Request processing**: Implemented `mcpStep` function that:
  - Decodes incoming JSON-RPC requests and validates their structure
  - Handles `initialize` method to transition from uninitialized to initialized state
  - Enforces initialization requirement for all other methods (returns -32002 error if not initialized)
  - Validates capability availability before dispatching methods (returns -32601 if capability absent)
  - Validates method parameters and returns -32602 for invalid params
  - Silently accepts notifications (messages without `id`) without state changes
  - Returns proper JSON-RPC responses with correct error codes
- **Error handling**: Integrated RPC error types (`invalidRequest`, `invalidParams`, `methodNotFound`) and added custom `notInitialized` error
- **Comprehensive test suite**: Added `proof.f.ts` with 15 test cases covering:
  - Lifecycle transitions and initialization
  - Parameter validation
  - Capability gating
  - Error responses for various invalid states and inputs
  - Tools method dispatch (`tools/list` and `tools/call`)

## Implementation Details
- The state machine is pure and functional—`mcpStep` returns both the new state and an effect
- Notifications are handled transparently without affecting session state
- Capability checking is performed before method dispatch to prevent unauthorized access
- All responses follow JSON-RPC 2.0 specification with proper error codes

https://claude.ai/code/session_01EFqwwrzYqJcxi2AGJNTdtR

### CODE REVIEW

**Comment by sergey-shandar** [comment]
@codex review

**Review by sergey-shandar — COMMENTED** [review]
## Review: MCP Session State Machine

Well-structured implementation that fits the project's functional/pure style. One behavioral issue worth resolving before merge, plus a few smaller points.

---

### Medium

**`initialize` can be called multiple times without error**
`mcpStep` handles `initialize` in any state, including after it has already been initialized. The second call silently overwrites the state and returns success. The MCP spec (2024-11-05) says `initialize` must only be sent once; re-sending it should be rejected with `-32600 Invalid Request`. A test case for this is also missing from `proof.f.ts`.

---

### Low

**`InitializedState` stores fields already in `McpConfig`**
`InitializedState` captures `capabilities` and `protocolVersion`, but these are already available in `config` which is passed to every `mcpStep` call. Unless the negotiated values can differ from the server config, this could be simplified to a marker type (e.g., `type InitializedState = true`), reducing redundancy.

**`_jsonrpc` duplicates an unexported constant from `fs/json/rpc/module.f.ts`**
`rpc/module.f.ts` already has `const jsonrpc = '2.0' as const` but it isn't exported. Exporting it from there and importing it here would avoid the duplication.

**`_okResponse` uses `as Unknown` cast**
```ts
({ jsonrpc: _jsonrpc, result: result as Unknown, id })
```
This bypasses the RTTI discipline used elsewhere in the codebase. A typed helper or proper schema encoding would be more consistent.

---

### Missing test cases

1. Double `initialize` call — would catch the re-initialization bug above.
2. `tools/call` with `arguments: null` or absent — edge case in `option(record(unknown))`.
3. Notification with an unknown method name — documents intended behavior (silent accept vs. error).

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by sergey-shandar** [comment]
@codex review

File: fs/json/module.f.ts [hunk 1]
```diff
@@ -33,7 +33,7 @@ export const primitive = or(null, rttiBoolean, rttiNumber, rttiString)
 export const unknown = () => ['or', primitive, object, array] as const
 
 /** rtti schema matching a JSON object: `{ readonly [k: string]: Unknown }`. */
-export const object = record(unknown)
+export const object = record(or(unknown, undefined))
```
[the remark this answers is not in the document]
Comment (sergey-shandar) [reply fs/json/module.f.ts]: The solution is to fix `Ts`.

**Comment by sergey-shandar** [comment]
@codex review

**Comment by sergey-shandar** [comment]
@codex review

**Comment by sergey-shandar** [comment]
@codex review

**Comment by sergey-shandar** [comment]
@codex review

**Comment by sergey-shandar** [comment]
@codex review

**Comment by sergey-shandar** [comment]
@codex review

**Comment by sergey-shandar** [comment]
@codex review

File: fs/types/ts/module.f.ts [hunk 2]
```diff
@@ -35,7 +35,7 @@ export const printer = (mut?: true): Printer => {
         struct: (fields: readonly (readonly[string, string])[]) =>
             structX(fields.map(([k, v]) => `${ro}${JSON.stringify(k)}:${v}`)),
         array: (type: string) => `${ro}(${type})[]`,
-        record: (type: string) => structX([`${ro}[k:string]:${type}`]),
+        record: (type: string) => structX([`${ro}[k:string]?:${type}`]),
```
[the remark this answers is not in the document]
Comment (sergey-shandar) [reply fs/types/ts/module.f.ts]: Right, we should print `[k in string]?:`.

**Comment by sergey-shandar** [comment]
@codex review

**Comment by sergey-shandar** [comment]
@claude review
